### ISSUE TITLE
title

### ISSUE THREAD

**Issue by hodaifayahia** [hodaifayahia/ECG#1]
# 🏥 ECG-Mustapha ML Enhancement: Medical Data Extraction & Machine Learning Pipeline

## 🎯 PROJECT GOAL
Enhance the ECG-Mustapha repository to extract ALL information from coronarography PDFs (patient demographics, clinical findings, coronary anatomy) and convert everything to numerical features for machine learning training. The system should handle both ECG signal analysis AND clinical report data extraction.

### 1. PATIENT DATA EXTRACTION FROM MEDICAL PDFs
Extract and structure data from French coronarography reports including:
- Patient demographics (name, age, gender, exam date)
- Clinical indications (cardiopathie ischémique, scintigraphie, scanner)
- Procedure details (technique, access point, catheter size)
- Coronary findings (trunk, IVA, diagonal, circumflex, right coronary)
- Stenosis locations and severity (serrée, modérée, occluse)
- Treatment decisions (pontage, angioplastie, medical)
- Diagnosis conclusions (tritronculaire, vessel counts)

### 2. NUMERICAL FEATURE ENGINEERING
Convert ALL text-based medical findings into numerical/categorical features:
- Binary features (yes/no): stenosis presence, occlusion, infiltration
- Ordinal features (0-3): stenosis severity (none/mild/moderate/severe)
- Categorical encoded: vessel locations, treatment types
- Computed scores: total stenosis score, vessel involvement, severity index

### 3. DATABASE SCHEMA FOR ML TRAINING
Create comprehensive PostgreSQL tables:

-- Patient clinical data table
CREATE TABLE patient_clinical_data (
id SERIAL PRIMARY KEY,
patient_id INTEGER REFERENCES patients(id),

text
-- Demographics
age INTEGER,
gender VARCHAR(10),
exam_date DATE,

-- Clinical Indication (binary)
indication_ischemic_heart_disease BOOLEAN,
indication_positive_scintigraphy BOOLEAN,
indication_positive_scanner BOOLEAN,

-- Procedure Details
access_point VARCHAR(20),  -- 'radial' or 'femoral'
catheter_size FLOAT,  -- 6.0, 7.0, etc.

-- LEFT CORONARY - Trunk
trunk_size INTEGER,  -- 1=small, 2=medium, 3=large
trunk_infiltration BOOLEAN,
trunk_stenosis BOOLEAN,
trunk_stenosis_severity INTEGER,  -- 0-3

-- LEFT CORONARY - IVA (LAD)
iva_developed BOOLEAN,
iva_atheroma BOOLEAN,
iva_stenosis_severity INTEGER,  -- 0-3
iva_stenosis_location VARCHAR(20),  -- 'proximal', 'middle', 'distal'
iva_stenosis_length VARCHAR(10),  -- 'short' or 'long'
iva_timi_flow INTEGER,  -- 1, 2, or 3

-- LEFT CORONARY - Diagonal
diagonal_present BOOLEAN,
diagonal_caliber INTEGER,  -- 1=small, 2=good, 3=large
diagonal_stenosis_severity INTEGER,  -- 0-3
diagonal_stenosis_location VARCHAR(20),

-- LEFT CORONARY - Circumflex
circumflex_atheroma BOOLEAN,
circumflex_stenosis_severity INTEGER,  -- 0-3
circumflex_stenosis_location VARCHAR(20),
circumflex_occlusion BOOLEAN,
circumflex_occlusion_location VARCHAR(20),
circumflex_collateral_network BOOLEAN,
circumflex_marginals_count INTEGER,

-- LEFT CORONARY - Bissectrice (if present)
bissectrice_present BOOLEAN,
bissectrice_stenosis_severity INTEGER,
bissectrice_stenosis_location VARCHAR(20),

-- RIGHT CORONARY
right_coronary_atheroma BOOLEAN,
right_coronary_stenosis_severity INTEGER,  -- 0-3
right_coronary_stenosis_length VARCHAR(10),
right_coronary_stenosis_location VARCHAR(20),
right_coronary_occlusion BOOLEAN,
right_coronary_occlusion_location VARCHAR(20),
right_coronary_collateral_network BOOLEAN,

-- COMPUTED SCORES
total_stenosis_score FLOAT,  -- sum of all stenosis severities
vessel_involvement_score INTEGER,  -- 0-3 (number of vessels affected)
occlusion_count INTEGER,  -- total number of occlusions
severity_index FLOAT,  -- weighted severity score

-- DIAGNOSIS (TARGET VARIABLES)
diagnosis_three_vessel_disease BOOLEAN,
diagnosis_vessels_affected_count INTEGER,
diagnosis_severity_level VARCHAR(20),  -- 'mild', 'moderate', 'severe'

-- TREATMENT DECISION (TARGET VARIABLES)
treatment_bypass_surgery BOOLEAN,
treatment_angioplasty BOOLEAN,
treatment_medical_only BOOLEAN,
treatment_adhoc BOOLEAN,
treatment_urgency_score INTEGER,  -- 0-3

-- Risk Factors (extract from medical history if present)
risk_hypertension BOOLEAN,
risk_diabetes BOOLEAN,
risk_smoking BOOLEAN,
risk_previous_mi BOOLEAN,
risk_family_history BOOLEAN,

created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);

-- Create indexes for ML queries
CREATE INDEX idx_clinical_severity ON patient_clinical_data(severity_index);
CREATE INDEX idx_clinical_treatment ON patient_clinical_data(treatment_bypass_surgery, treatment_angioplasty);
CREATE INDEX idx_clinical_vessels ON patient_clinical_data(vessel_involvement_score);

text

### 4. PYTHON MODULE: PDF TEXT EXTRACTION & NLP

Create `ecg_modules/medical_text_extractor.py`:

import fitz # PyMuPDF
import re
from datetime import datetime

class MedicalReportExtractor:
"""Extract structured data from French coronarography reports"""

text
def __init__(self):
    # Regex patterns for French medical terminology
    self.patterns = {
        'patient_name': r'Nom\s*:\s*([A-ZÀ-Ü]+).*?Prénom\s*:\s*([A-ZÀ-Üa-zà-ü]+)',
        'age': r'Age\s*:\s*(\d+)',
        'exam_date': r'Date.*?:\s*(\d{2}-\d{2}-\d{4})',
        'indication': r'Indication\s*:\s*(.+?)\.', 
        'catheter_size': r'(\d)F',
        'timi_flow': r"TIMI\s*['\"]([I]{1,3})['\"]",
        'stenosis_severe': r'sténose\s+(?:très\s+)?serrée',
        'stenosis_moderate': r'sténose\s+modérée',
        'occlusion': r'occlu[se]',
        'tritronculaire': r'tritronculaire',
        'bypass': r'[Pp]ontage',
        'angioplasty': r'[Aa]ngioplastie'
    }

def extract_from_pdf(self, pdf_path_or_bytes):
    """Extract all text and structured data from PDF"""
    # Open PDF
    if isinstance(pdf_path_or_bytes, bytes):
        doc = fitz.open(stream=pdf_path_or_bytes, filetype="pdf")
    else:
        doc = fitz.open(pdf_path_or_bytes)
    
    # Extract full text
    full_text = ""
    for page in doc:
        full_text += page.get_text()
    
    doc.close()
    
    # Parse structured data
    structured_data = self._parse_medical_report(full_text)
    
    return full_text, structured_data

def _parse_medical_report(self, text):
    """Parse French medical report text into structured data"""
    data = {}
    
    # Extract patient demographics
    name_match = re.search(self.patterns['patient_name'], text, re.IGNORECASE)
    if name_match:
        data['last_name'] = name_match.group(1)
        data['first_name'] = name_match.group(2)
    
    age_match = re.search(self.patterns['age'], text)
    if age_match:
        data['age'] = int(age_match.group(1))
    
    # Extract clinical findings
    data['stenosis_present'] = bool(re.search(self.patterns['stenosis_severe'], text, re.IGNORECASE))
    data['occlusion_present'] = bool(re.search(self.patterns['occlusion'], text, re.IGNORECASE))
    data['three_vessel_disease'] = bool(re.search(self.patterns['tritronculaire'], text, re.IGNORECASE))
    
    # Extract treatment decision
    data['requires_bypass'] = bool(re.search(self.patterns['bypass'], text, re.IGNORECASE))
    data['requires_angioplasty'] = bool(re.search(self.patterns['angioplasty'], text, re.IGNORECASE))
    
    # Extract TIMI flow
    timi_match = re.search(self.patterns['timi_flow'], text)
    if timi_match:
        timi_roman = timi_match.group(1)
        data['timi_flow'] = len(timi_roman)  # I=1, II=2, III=3
    
    # Count stenoses and occlusions
    data['stenosis_count'] = len(re.findall(self.patterns['stenosis_severe'], text, re.IGNORECASE))
    data['occlusion_count'] = len(re.findall(self.patterns['occlusion'], text, re.IGNORECASE))
    
    # Parse coronary sections
    data['left_coronary'] = self._parse_coronary_section(text, 'gauche')
    data['right_coronary'] = self._parse_coronary_section(text, 'droite')
    
    # Calculate severity scores
    data['total_stenosis_score'] = self._calculate_stenosis_score(data)
    data['severity_level'] = self._determine_severity_level(data)
    
    return data

def _parse_coronary_section(self, text, side):
    """Parse left or right coronary findings"""
    # Implementation for detailed coronary parsing
    coronary_data = {}
    # Add parsing logic for each vessel
    return coronary_data

def _calculate_stenosis_score(self, data):
    """Calculate numerical severity score"""
    score = 0
    score += data.get('stenosis_count', 0) * 2
    score += data.get('occlusion_count', 0) * 3
    if data.get('three_vessel_disease'):
        score += 5
    return score

def _determine_severity_level(self, data):
    """Determine severity level based on findings"""
    score = data.get('total_stenosis_score', 0)
    if score >= 10:
        return 'severe'
    elif score >= 5:
        return 'moderate'
    else:
        return 'mild'
text

### 5. FLASK ROUTES FOR CLINICAL DATA

Add to `app.py`:

from ecg_modules.medical_text_extractor import MedicalReportExtractor

medical_extractor = MedicalReportExtractor()

@app.route("/api/upload_clinical_report", methods=["POST"])
def upload_clinical_report():
"""Upload coronarography PDF and extract clinical data"""
try:
file = request.files["file"]
patient_id = request.form.get('patient_id')

text
    # Extract data from PDF
    pdf_bytes = file.read()
    full_text, structured_data = medical_extractor.extract_from_pdf(pdf_bytes)
    
    # Save to database
    db.save_clinical_data(patient_id, structured_data, full_text)
    
    return jsonify({
        "status": "success",
        "extracted_data": structured_data
    })
    
except Exception as e:
    return jsonify({"error": str(e)}), 500
@app.route("/api/export_ml_complete_dataset", methods=["GET"])
def export_ml_complete_dataset():
"""Export complete dataset: ECG metrics + Clinical data"""
try:
# Query that joins ECG analyses with clinical data
query = """
SELECT
p.patient_id,
p.age,
p.gender,
c.*, -- All clinical features
e.heart_rate_bpm,
e.sdnn_ms,
e.rmssd_ms,
e.lf_hf_ratio,
e.raw_metrics
FROM patient_clinical_data c
LEFT JOIN patients p ON c.patient_id = p.id
LEFT JOIN ecg_analyses e ON p.id = e.patient_id
WHERE c.severity_index IS NOT NULL
"""

text
    records = db.execute_query(query)
    
    # Export as CSV
    return create_csv_response(records)
    
except Exception as e:
    return jsonify({"error": str(e)}), 500
text

### 6. MACHINE LEARNING PIPELINE

Create `ml_training/train_model.py`:

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, confusion_matrix

class CoronaryDiseasePredictor:
"""ML model for predicting coronary disease severity and treatment"""

text
def __init__(self):
    self.scaler = StandardScaler()
    self.model = None
    
def prepare_features(self, df):
    """Prepare numerical features for ML"""
    
    # Numerical features
    numerical_features = [
        'age',
        'catheter_size',
        'trunk_size',
        'trunk_infiltration',
        'iva_stenosis_severity',
        'iva_timi_flow',
        'diagonal_stenosis_severity',
        'circumflex_stenosis_severity',
        'circumflex_occlusion',
        'circumflex_marginals_count',
        'right_coronary_stenosis_severity',
        'right_coronary_occlusion',
        'total_stenosis_score',
        'vessel_involvement_score',
        'occlusion_count',
        'severity_index',
        # ECG features
        'heart_rate_bpm',
        'sdnn_ms',
        'rmssd_ms',
        'lf_hf_ratio'
    ]
    
    # Categorical features (one-hot encode)
    categorical_features = [
        'gender',
        'access_point',
        'iva_stenosis_location',
        'circumflex_stenosis_location'
    ]
    
    # Extract features
    X_numerical = df[numerical_features].fillna(0)
    X_categorical = pd.get_dummies(df[categorical_features], drop_first=True)
    
    X = pd.concat([X_numerical, X_categorical], axis=1)
    
    return X

def train_bypass_predictor(self, df):
    """Train model to predict bypass surgery requirement"""
    X = self.prepare_features(df)
    y = df['treatment_bypass_surgery']
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    
    # Scale features
    X_train_scaled = self.scaler.fit_transform(X_train)
    X_test_scaled = self.scaler.transform(X_test)
    
    # Train model
    self.model = GradientBoostingClassifier(n_estimators=100, random_state=42)
    self.model.fit(X_train_scaled, y_train)
    
    # Evaluate
    y_pred = self.model.predict(X_test_scaled)
    print("Bypass Surgery Prediction:")
    print(classification_report(y_test, y_pred))
    
    return self.model

def train_severity_classifier(self, df):
    """Train model to classify disease severity"""
    X = self.prepare_features(df)
    y = df['diagnosis_severity_level']
    
    # Multi-class classification
    # Implementation similar to above
    pass
text

### 7. FRONTEND UI FOR CLINICAL DATA UPLOAD

Add to `static/index.html`:

 <div class="section" id="clinical-section"> <h2>📋 Clinical Report Upload</h2>
text
<div class="card">
    <h3>Upload Coronarography PDF</h3>
    <p>Upload French coronarography reports - data will be automatically extracted</p>
    
    <input type="text" id="clinical-patient-id" placeholder="Patient ID">
    <input type="file" id="clinical-pdf" accept=".pdf">
    <button onclick="uploadClinicalReport()">Extract & Upload Clinical Data</button>
    
    <div id="extracted-data-preview"></div>
</div>

<div class="card">
    <h3>📊 ML Dataset Export</h3>
    <p>Export complete dataset: ECG metrics + Clinical findings</p>
    
    <button onclick="exportCompleteMLDataset()" class="btn-primary">
        Download ML Training Dataset (CSV)
    </button>
</div>
</div> <script> async function uploadClinicalReport() { const patientId = document.getElementById('clinical-patient-id').value; const file = document.getElementById('clinical-pdf').files; const formData = new FormData(); formData.append('file', file); formData.append('patient_id', patientId); const response = await fetch('/api/upload_clinical_report', { method: 'POST', body: formData }); const result = await response.json(); // Display extracted data document.getElementById('extracted-data-preview').innerHTML = ` <h4>Extracted Data:</h4> <pre>${JSON.stringify(result.extracted_data, null, 2)}</pre>
text
`;
}

async function exportCompleteMLDataset() {
window.location.href = '/api/export_ml_complete_dataset';
}
</script>

text

## 🎯 IMPLEMENTATION CHECKLIST

- [ ] Create database migration with `patient_clinical_data` table
- [ ] Implement `MedicalReportExtractor` class with French NLP
- [ ] Add regex patterns for all coronary findings
- [ ] Implement numerical encoding for all categorical features
- [ ] Create severity score calculation algorithms
- [ ] Add Flask routes for clinical report upload
- [ ] Build ML export endpoint combining ECG + Clinical data
- [ ] Create frontend UI for clinical data upload
- [ ] Implement ML training pipeline
- [ ] Add data validation and error handling
- [ ] Create comprehensive test suite with sample PDFs
- [ ] Add documentation with feature descriptions

## 📊 EXPECTED OUTPUT

Final CSV for ML training should contain ~70+ columns:
patient_id, age, gender,
trunk_size, trunk_infiltration, trunk_stenosis_severity,
iva_stenosis_severity, iva_timi_flow, iva_atheroma,
diagonal_stenosis_severity, circumflex_stenosis_severity,
circumflex_occlusion, right_coronary_stenosis_severity,
total_stenosis_score, vessel_involvement_score, severity_index,
heart_rate_bpm, sdnn_ms, rmssd_ms, lf_hf_ratio,
diagnosis_three_vessel_disease, treatment_bypass_surgery,
treatment_angioplasty, diagnosis_severity_level

text

## 🔥 KEY FEATURES

✅ Extract ALL information from French medical PDFs
✅ Convert text to 50+ numerical/categorical features
✅ Combine clinical data + ECG signal analysis
✅ Multiple target variables for different ML tasks
✅ Severity scoring algorithms
✅ Ready-to-use ML training pipeline
✅ Complete database schema with indexes
✅ Production-ready Flask API

@copilot Please implement this complete system!
